### ISSUE TITLE
BooleanPropertyNaming reports double for local variables (in some TBD case)

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5392]
## Context
Consider code like this (note the implicit types of all `isEmpty()` are `Boolean`):
```kotlin
val feed2: Feed = generateSequence { fixture.build<Feed>() }
	.first { feed2 ->
		val uniqueFilms = feed2.films.isEmpty()
		val uniqueFilms2 =
			feed2.films.map(Feed.Film::id).isEmpty()
		val uniqueAttributes =
			feed2.attributes.map(Feed.Attribute::code)
				.isEmpty()
		uniqueFilms && uniqueFilms2 && uniqueAttributes
	}
```

## Observed Behavior
Notice the line numbers: 18, 19, 21, 18, 19, 21
```
> Task :backend:feed:detektTest FAILED
naming - 30min debt
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:18:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:19:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:21:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:18:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:19:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:21:5
```

## Expected Behavior
1 report for each local variable.
```
> Task :backend:feed:detektTest FAILED
naming - 30min debt
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:18:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:19:5
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at backend\feed\src\test\kotlin\net\twisterrob\cinema\cineworld\sync\syndication\FeedTest.kt:21:5
```

## Steps to Reproduce
Couldn't figure a simple way, so:
 * Clone https://github.com/TWiStErRob/net.twisterrob.cinema
 * Check out commit db7342d5a0895f93fa595568883165c24f25dc07
 * Run `gradlew :backend:feed:detektTest`
 * Notice that only `BooleanPropertyNaming` is duplicated, the others are unique

## Your Environment

* Version of detekt used: 1.21.0
* Version of Gradle used (if applicable): 7.5.1
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: Windows
* Link to your project (if it's a public repository): https://github.com/TWiStErRob/net.twisterrob.cinema/blob/db7342d5a0895f93fa595568883165c24f25dc07/Heroku/backend/feed/src/test/kotlin/net/twisterrob/cinema/cineworld/sync/syndication/FeedTest.kt

**Comment by schalkms** [comment]
Putting this code in our test environment doesn't report the same code location twice.
@TWiStErRob do you run detekt in parallel mode?

**Comment by TWiStErRob** [comment]
Yes, but I tried (=false) and it doesn't make a difference: https://github.com/TWiStErRob/net.twisterrob.cinema/blob/db7342d5a0895f93fa595568883165c24f25dc07/Heroku/gradle/plugins/src/main/kotlin/net/twisterrob/cinema/heroku/plugins/DetektPlugin.kt#L27

**Comment by TWiStErRob** [comment]
I'm trying to narrow it down:
 * my repro in #5391 also failed, it only show 3 warnings, BUT if I copy that code into the same project and do the repro above, it shows 6 findings.
 * I tried removing all sources of from my test sourceSet, same behavior.
 * I tried removing all sources of from my main sourceSet, same behavior.
 * I tried removing all dependencies and other modules of from my Gradle projects, same behavior.
 * hint: if I remove the `val feed2: Feed =` part, it reduces to 3 findings. (if I just rename it, it stays); I cannot conclude anything from this
 * Removed kapt and java plugins, same behavior.
 * Removed all other mentions of detekt in configuration and inlined my plugin... same behavior
 * Moved it to the main source set, same behavior.
 * I removed all other build gradle code, same behavior.
 * I moved code to `:` submodule, same behavior.
 * I removed all extra detekt config (except allrules = true), same behavior.

... getting there, but so far it's a big wot?

**Comment by TWiStErRob** [comment]
... continuing
 * removed all detekt config except `config = files("detekt.yml")`, same behavior.
 * removed package, same behavior.
 * removed all Gradle properties, same behavior.

This is the point where I cannot think of any more simplifications, and no conclusive cause found. Creating minimalish self-contained repro:
 * Clone https://github.com/TWiStErRob/repros/tree/master/detekt/BooleanPropertyNaming-duplicate-findings
 * `gradlew :detektMain`

Note: I have enabled `UseDataClass` as well for control, as it doesn't duplicate findings.

```
> Task :detektMain FAILED
. root project> 0% CONFIGURING [86ms]

1 kotlin file was analyzed.
...\src\main\kotlin\f.kt:11:4: Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms [BooleanPropertyNaming]
...\src\main\kotlin\f.kt:12:4: Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms2 [BooleanPropertyNaming]
...\src\main\kotlin\f.kt:14:4: Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueAttributes [BooleanPropertyNaming]
...\src\main\kotlin\f.kt:11:4: Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms [BooleanPropertyNaming]
...\src\main\kotlin\f.kt:12:4: Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueFilms2 [BooleanPropertyNaming]
...\src\main\kotlin\f.kt:14:4: Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqueAttributes [BooleanPropertyNaming]
...\src\main\kotlin\f.kt:1:1: The class Feed defines no functionality and only holds data. Consider converting it to a data class. [UseDataClass]
...\src\main\kotlin\f.kt:5:1: The class Film defines no functionality and only holds data. Consider converting it to a data class. [UseDataClass]
...\src\main\kotlin\f.kt:6:1: The class Attribute defines no functionality and only holds data. Consider converting it to a data class. [UseDataClass]

...\src\main\kotlin\f.kt - 45min debt
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:11:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:12:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:14:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:11:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:12:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:14:4
        UseDataClass - [The class Feed defines no functionality and only holds data. Consider converting(...)] at ...\src\main\kotlin\f.kt:1:1
        UseDataClass - [The class Film defines no functionality and only holds data. Consider converting(...)] at ...\src\main\kotlin\f.kt:5:1
        UseDataClass - [The class Attribute defines no functionality and only holds data. Consider conve(...)] at ...\src\main\kotlin\f.kt:6:1

Overall debt: 45min

naming - 30min debt
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:11:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:12:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:14:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:11:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:12:4
        BooleanPropertyNaming - [Boolean property name should match a ^(is|has|are) pattern. Actual name is uniqu(...)] at ...\src\main\kotlin\f.kt:14:4
style - 15min debt
        UseDataClass - [The class Feed defines no functionality and only holds data. Consider converting(...)] at ...\src\main\kotlin\f.kt:1:1
        UseDataClass - [The class Film defines no functionality and only holds data. Consider converting(...)] at ...\src\main\kotlin\f.kt:5:1
        UseDataClass - [The class Attribute defines no functionality and only holds data. Consider conve(...)] at ...\src\main\kotlin\f.kt:6:1

Overall debt: 45min

Complexity Report:
        - 20 lines of code (loc)
        - 19 source lines of code (sloc)
        - 16 logical lines of code (lloc)
        - 0 comment lines of code (cloc)
        - 3 cyclomatic complexity (mcc)
        - 1 cognitive complexity
        - 9 number of total code smells
        - 0% comment source ratio
        - 187 mcc per 1,000 lloc
        - 562 code smells per 1,000 lloc

Project Statistics:
        - number of properties: 4
        - number of functions: 1
        - number of classes: 3
        - number of packages: 1
        - number of kt files: 1

Successfully generated Checkstyle XML report at ...\build\reports\detekt\main.xml
Successfully generated plain text report at ...\build\reports\detekt\main.txt
Successfully generated HTML report at ...\build\reports\detekt\main.html
Successfully generated SARIF: a standard format for the output of static analysis tools at ...\build\reports\detekt\main.sarif
Successfully generated Markdown report at ...\build\reports\detekt\main.md

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':detektMain'.
> Analysis failed with 9 weighted issues.
```

**Comment by TWiStErRob** [comment]
Based on the repro I did a `gradlew detektMain -Dorg.gradle.debug=true --no-daemon` with a breakpoint on `report()` in `BooleanPropertyNaming` and it seems the duplication comes from the `super.visitProperty` + `validateDeclaration` call, because there's that extra `val feed2: Feed` variable (as seen in hint above). This is the point where my PSI knowledge lacks: I cannot see why for the outer declaration it would still find the inner violations. @schalkms please pick up from here.

**Comment by TWiStErRob** [comment]
oh, no-repro on rc2

**Comment by TWiStErRob** [comment]
Fixed by https://github.com/detekt/detekt/pull/5212 which was first available in [v1.22.0-RC1](https://github.com/detekt/detekt/releases/tag/v1.22.0-RC1)
